src/metadata_map.rs

```diff
@@ -0,0 +1,1716 @@
+    /// assert_eq!(map.get("x-hello").unwrap(), "1");
+    /// ```
+    pub fn entry<K>(&mut self, key: K) -> Result<Entry<MetadataValue>, InvalidMetadataKey>
+        where K: AsMetadataKey
+    {
+        match key.entry(self) {
+            Ok(entry) => {
+                Ok(match entry {
+                    http::header::Entry::Occupied(e) =>
+                        Entry::Occupied(OccupiedEntry { inner: e }),
+                    http::header::Entry::Vacant(e) =>
+                        Entry::Vacant(VacantEntry { inner: e }),
+                })
+            }
+            Err(_) => Err(InvalidMetadataKey::new()),
+        }
+    }
+
+    /// Inserts a key-value pair into the map.
+    ///
+    /// If the map did not previously have this key present, then `None` is
+    /// returned.
+    ///
+    /// If the map did have this key present, the new value is associated with
+    /// the key and all previous values are removed. **Note** that only a single
+    /// one of the previous values is returned. If there are multiple values
+    /// that have been previously associated with the key, then the first one is
+    /// returned. See `insert_mult` on `OccupiedEntry` for an API that returns
+    /// all values.
+    ///
+    /// The key is not updated, though; this matters for types that can be `==`
+    /// without being identical.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use tower_grpc::metadata::*;
+    /// let mut map = MetadataMap::new();
+    /// assert!(map.insert("x-host", "world".parse().unwrap()).is_none());
+    /// assert!(!map.is_empty());
+    ///
+    /// let mut prev = map.insert("x-host", "earth".parse().unwrap()).unwrap();
+    /// assert_eq!("world", prev);
+    /// ```
+    pub fn insert<K>(&mut self, key: K, val: MetadataValue) -> Option<MetadataValue>
+        where K: IntoMetadataKey
+    {
+        key.insert(self, val)
+    }
+
+    /// Inserts a key-value pair into the map.
+    ///
+    /// If the map did not previously have this key present, then `false` is
+    /// returned.
+    ///
+    /// If the map did have this key present, the new value is pushed to the end
+    /// of the list of values currently associated with the key. The key is not
+    /// updated, though; this matters for types that can be `==` without being
+    /// identical.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use tower_grpc::metadata::*;
+    /// let mut map = MetadataMap::new();
+    /// assert!(map.insert("x-host", "world".parse().unwrap()).is_none());
+    /// assert!(!map.is_empty());
+    ///
+    /// map.append("x-host", "earth".parse().unwrap());
+    ///
+    /// let values = map.get_all("x-host");
+    /// let mut i = values.iter();
+    /// assert_eq!("world", *i.next().unwrap());
+    /// assert_eq!("earth", *i.next().unwrap());
+    /// ```
+    pub fn append<K>(&mut self, key: K, value: MetadataValue) -> bool
+        where K: IntoMetadataKey
+    {
+        key.append(self, value)
```

Comment:
Can this be implemented in terms of HeaderMap::append?

Reply:
See above